File: Interview Preparation Kit/Stacks and Queues/BalancedBrackets.java
// https://www.hackerrank.com/challenges/balanced-brackets/problem
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {

    // Complete the isBalanced function below.
    static String isBalanced(String s) {
        Stack<Character> stack = new Stack<>();

        int s_len = s.length();
        for (int i = 0; i < s_len; i++) {
            char bracket = s.charAt(i);
            if (bracket == '(' || bracket == '{' || bracket == '[') {
                stack.push(bracket);
            }
            else {
                if (stack.isEmpty()) {
                    return "NO";
                }

                char top = stack.peek();
                if (bracket == ')' && top == '(') {
                    stack.pop();
                }
                else if (bracket == '}' && top == '{') {
                    stack.pop();
                }
                else if (bracket == ']' && top == '[') {
                    stack.pop();
                }
                else {
                    return "NO";
                }
            }
        }

        if (stack.isEmpty()) {
            return "YES";
        }

        return "NO";
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int t = scanner.nextInt();
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        for (int tItr = 0; tItr < t; tItr++) {
            String s = scanner.nextLine();

            String result = isBalanced(s);

            bufferedWriter.write(result);
            bufferedWriter.newLine();
        }

        bufferedWriter.close();

        scanner.close();
    }
}

File: Interview Preparation Kit/Stacks and Queues/BalancedBrackets.cpp
// https://www.hackerrank.com/challenges/balanced-brackets/problem
#include <bits/stdc++.h>

using namespace std;

// Complete the isBalanced function below.
string isBalanced(string s) {
    stack<char> brackets;

    int s_len = (int)s.length();
    for (int i = 0; i < s_len; i++) {
        char bracket = s[i];
        if (bracket == '(' || bracket == '{' || bracket == '[') {
            brackets.push(bracket);
        }
        else {
            if (brackets.empty()) {
                return "NO";
            }

            char top = brackets.top();
            if (bracket == ')' && top == '(') {
                brackets.pop();
            }
            else if (bracket == '}' && top == '{') {
                brackets.pop();
            }
            else if (bracket == ']' && top == '[') {
                brackets.pop();
            }
            else {
                return "NO";
            }
        }
    }

    if (brackets.empty()) {
        return "YES";
    }

    return "NO";
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    int t;
    cin >> t;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    for (int t_itr = 0; t_itr < t; t_itr++) {
        string s;
        getline(cin, s);

        string result = isBalanced(s);

        fout << result << "\n";
    }

    fout.close();

    return 0;
}
